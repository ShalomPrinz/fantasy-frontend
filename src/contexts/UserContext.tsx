import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";

import { toast } from "react-toastify";

import {
  getUserInfo,
  loadIdToken,
  loginUser,
  logoutUser,
  registerUser,
  removeIdToken,
  saveIdToken,
} from "../services";
import { LoginUser, RegisterUser, Team, User } from "../types";

enum State {
  LOADING_USER,
  LOGGED_USER,
  NO_LOGGED_USER,
}

interface UserContextValue {
  loading: boolean;
  login: (info: {}) => Promise<void>;
  logout: () => Promise<void>;
  register: (info: {}) => Promise<void>;
  state: State;
  user: User | undefined;
}

const UserContext = createContext<UserContextValue | undefined>(undefined);

function useUser() {
  const context = useContext(UserContext);
  if (typeof context === "undefined") {
    throw new Error("useUser must be used within a UserProvider");
  }
  return context;
}

interface UserProviderProps {
  children: ReactNode;
}

function UserProvider({ children }: UserProviderProps) {
  const [loading, setLoading] = useState(true);
  const [currentUser, setCurrentUser] = useState<User | undefined>();
  const state =
    typeof currentUser === "undefined"
      ? loading
        ? State.LOADING_USER
        : State.NO_LOGGED_USER
      : State.LOGGED_USER;

  const loadUser = async () => {
    if (loadIdToken()) {
      setLoading(true);
      const res = await getUserInfo().catch(({ response }) => {
        if (response.status === 401) {
          toast.error(
            "Your saved login info is outdated. Please Log In again",
            {
              toastId: "outdated",
            }
          );
          removeIdToken();
        }
      });
      if (res?.data?.user) {
        const { leagues, nickname, team } = res.data.user;
        const userTeam = new Team(team, true);
        const user = new User(leagues, nickname, userTeam);
        setCurrentUser(user);
      }
    }
    setLoading(false);
  };

  useEffect(() => {
    loadUser();
  }, []);

  const login = async (loginInfo: LoginUser) => {
    const idToken = await loginUser(loginInfo);
    saveIdToken(idToken);
    loadUser();
  };

  const logout = async () => {
    removeIdToken();
    setCurrentUser(undefined);
    await logoutUser();
  };

  const register = async (registerInfo: RegisterUser) => {
    const idToken = await registerUser(registerInfo);
    saveIdToken(idToken);
    loadUser();
  };

  const value: UserContextValue = {
    loading,
    login,
    logout,
    register,
    state,
    user: currentUser,
  };

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
}

export { State, useUser };
export default UserProvider;
