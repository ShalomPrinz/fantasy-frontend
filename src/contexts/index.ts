import TeamProvider from "./TeamContext";
import UserProvider from "./UserContext";

export { useTeamState, useTeamUpdate } from "./TeamContext";
export { useUser } from "./UserContext";
export { TeamProvider, UserProvider };
