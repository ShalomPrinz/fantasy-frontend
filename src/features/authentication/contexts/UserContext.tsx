import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";

import { toast } from "react-toastify";

import type { User } from "../../../types";
import {
  getUserInfo,
  loadIdToken,
  loginUser,
  logoutUser,
  registerUser,
  removeIdToken,
  saveIdToken,
} from "../services";
import { parseUser } from "../types";
import type { LoginUser, RegisterUser } from "../types";

enum UserState {
  LOADING_USER,
  LOGGED_USER,
  NO_LOGGED_USER,
}

interface UserContextValue {
  invalidate: () => Promise<void>;
  loading: boolean;
  login: (info: {}) => Promise<void>;
  logout: () => Promise<void>;
  register: (info: {}) => Promise<void>;
  state: UserState;
  user: User | undefined;
}

const UserContext = createContext<UserContextValue | undefined>(undefined);

function useUser() {
  const context = useContext(UserContext);
  if (typeof context === "undefined") {
    throw new Error("useUser must be used within a UserProvider");
  }
  return context;
}

interface UserProviderProps {
  children: ReactNode;
}

function UserProvider({ children }: UserProviderProps) {
  const [loading, setLoading] = useState(true);
  const [currentUser, setCurrentUser] = useState<User | undefined>();
  const state =
    typeof currentUser === "undefined"
      ? loading
        ? UserState.LOADING_USER
        : UserState.NO_LOGGED_USER
      : UserState.LOGGED_USER;

  const loadUser = async () => {
    if (loadIdToken()) {
      setLoading(true);
      const res = await getUserInfo().catch(({ response }) => {
        if (response.status === 401) {
          toast.error(
            "Your saved login info is outdated. Please Log In again",
            {
              toastId: "outdated",
            }
          );
          removeIdToken();
        }
      });
      if (res?.data?.user) {
        const user = parseUser(res.data.user);
        setCurrentUser(user);
      }
    }
    setLoading(false);
  };

  useEffect(() => {
    loadUser();
  }, []);

  const login = async (loginInfo: LoginUser) => {
    const idToken = await loginUser(loginInfo);
    saveIdToken(idToken);
    loadUser();
  };

  const logout = async () => {
    removeIdToken();
    setCurrentUser(undefined);
    await logoutUser();
  };

  const register = async (registerInfo: RegisterUser) => {
    const idToken = await registerUser(registerInfo);
    saveIdToken(idToken);
    loadUser();
  };

  const value: UserContextValue = {
    invalidate: loadUser,
    loading,
    login,
    logout,
    register,
    state,
    user: currentUser,
  };

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
}

export { UserProvider, UserState, useUser };
