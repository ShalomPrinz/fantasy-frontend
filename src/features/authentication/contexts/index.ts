export { UserProvider, UserState, useUser } from "./UserContext";
