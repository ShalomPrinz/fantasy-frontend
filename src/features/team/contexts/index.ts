import TeamProvider from "./TeamContext";

export { useTeamState, useTeamUpdate } from "./TeamContext";
export { TeamProvider };
