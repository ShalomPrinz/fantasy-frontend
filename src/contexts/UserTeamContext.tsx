import { createContext, ReactNode, useContext, useReducer } from "react";

import { Player, Team } from "../types";
import { useUser } from "./UserContext";

type Action = { type: "addPlayer"; payload: { player: Player } };
type Dispatch = (action: Action) => void;

const TeamContextState = createContext<Team | undefined>(undefined);
function useTeamState() {
  const team = useContext(TeamContextState);
  if (team === undefined)
    throw new Error("useTeamState must be used within a TeamProvider");
  return team;
}

const TeamContextUpdate = createContext<Dispatch | undefined>(undefined);
function useTeamUpdate() {
  const dispatch = useContext(TeamContextUpdate);
  if (dispatch === undefined)
    throw new Error("useTeamUpdate must be used within a TeamProvider");
  return (player: Player) =>
    dispatch({ type: "addPlayer", payload: { player } });
}

interface TeamProviderProps {
  children: ReactNode;
}

function TeamProvider({ children }: TeamProviderProps) {
  const { user } = useUser();
  const [team, dispatch] = useReducer(teamReducer, user!.team);

  return (
    <TeamContextState.Provider value={team}>
      <TeamContextUpdate.Provider value={dispatch}>
        {children}
      </TeamContextUpdate.Provider>
    </TeamContextState.Provider>
  );
}

function teamReducer(team: Team, action: Action) {
  switch (action.type) {
    case "addPlayer": {
      const { player } = action.payload;
      if (!team.contains(player)) {
        team.addPlayer(player);
      }
      return team.clone();
    }
  }
}

export default TeamProvider;
export { useTeamState, useTeamUpdate };
