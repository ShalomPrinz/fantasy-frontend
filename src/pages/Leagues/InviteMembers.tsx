import { useState } from "react";

import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import { useNavigate } from "react-router-dom";
import { toast } from "react-toastify";

import { Message, Search, Table } from "../../components";
import { UserState, useUser } from "../../contexts";
import { useLeagueInfo } from "../../hooks";
import { inviteLeagueMember, queryUsers } from "../../services";
import { QueriedUser } from "../../types";

interface InviteMembersProps {
  userId: string;
}

const InviteMembers = ({ userId }: InviteMembersProps) => {
  const { league, isLoading } = useLeagueInfo(userId);
  const { inviteMember, isInviting } = useInviteMembers(
    league?.id,
    league?.name
  );
  const navigate = useNavigate();
  const [users, setUsers] = useState([]);

  function handleQuery(query: string) {
    queryUsers(query).then((res) => setUsers(res.data.users || []));
  }

  if (isLoading)
    return <Message color="info" text="Loading League Information..." />;
  if (typeof league === "undefined")
    return <Message color="danger" text="You are not member of this league." />;

  const inviteTableColumns = [
    {
      id: 0,
      label: "Username",
      path: "username",
    },
    {
      id: 1,
      label: "Invite",
      content: ({ id, username }: QueriedUser) => {
        return (
          <button
            className="fs-4 bg-default rounded py-2 px-3"
            disabled={isInviting}
            onClick={() => inviteMember(id, username)}
            type="button"
          >
            Invite
          </button>
        );
      },
    },
  ];

  return (
    <main className="container mt-5 text-center">
      <Row>
        <h1 className="mb-5">Invite new members to {league!.name}</h1>
      </Row>
      <Row>
        <Col sm="3" className="mx-auto">
          <h3 className="m-4">Members: {league!.members.length}</h3>
          <button
            className="fs-2 bg-default rounded py-3 px-4 button-border-focus"
            onClick={() => navigate(-1)}
            type="button"
          >
            {league!.name}
          </button>
        </Col>
        <Col sm="6" className="mx-auto">
          <Search onChange={handleQuery} />
          <div className="w-75 mx-auto mt-4 fs-3">
            <Table columns={inviteTableColumns} data={users} />
            {!users.length && "No Users Found"}
          </div>
        </Col>
      </Row>
    </main>
  );
};

function useInviteMembers(
  leagueId: string | undefined,
  leagueName: string | undefined
) {
  const [isInviting, setIsInviting] = useState(false);

  function inviteMember(userId: string, username: string) {
    if (!leagueId || isInviting) return;

    setIsInviting(true);
    inviteLeagueMember(userId, leagueId)
      .then(() => {
        toast.success(`Successfully invited ${username} to ${leagueName}`);
      })
      .finally(() => {
        setIsInviting(false);
      });
  }

  return {
    inviteMember,
    isInviting,
  };
}

const InviteMembersWrapper = () => {
  const { state, user } = useUser();

  switch (state) {
    case UserState.LOADING_USER:
      return <Message color="info" text="Loading..." />;

    case UserState.LOGGED_USER:
      return <InviteMembers userId={user!.id} />;

    case UserState.NO_LOGGED_USER:
      return (
        <Message
          color="danger"
          text="Please Log In to invite members to your league"
        />
      );
  }
};

export default InviteMembersWrapper;
