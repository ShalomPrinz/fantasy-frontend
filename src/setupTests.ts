// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import axios from "axios";
import { setupServer } from "msw/node";

import { handlers } from "../test";
import * as TeamContext from "./contexts/TeamContext";
import * as UserContext from "./features/authentication/contexts/UserContext";
import { LeagueInfo, Message, Player, Team, User } from "./types";

jest.setTimeout(10000);

axios.defaults.baseURL = "";

const server = setupServer(...handlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

interface MockedUserContext {
  logout?: () => Promise<void>;
  state?: UserContext.UserState;
  user?: User;
}

interface MockedUser {
  inbox?: Message[];
  leagues?: LeagueInfo[];
  name?: string;
  players?: Player[];
}

export const getTestUser = (user?: MockedUser): User => ({
  id: "id",
  inbox: user?.inbox || [],
  leagues: user?.leagues || [],
  name: user?.name || "",
  team: new Team(user?.players || []),
});

export const mockUser = (context: MockedUserContext) => {
  const userContext = jest.spyOn(UserContext, "useUser");

  const { logout, state, user } = context;
  const mockedLogout = typeof logout !== "undefined" ? logout : async () => {};
  const mockedState =
    typeof state !== "undefined" ? state : UserContext.UserState.LOGGED_USER;
  const mockedUser = typeof user !== "undefined" ? user : getTestUser();

  userContext.mockImplementation(() => ({
    loading: false,
    login: async () => {},
    logout: mockedLogout,
    register: async () => {},
    state: mockedState,
    user: mockedUser,
    invalidate: async () => {},
  }));

  return userContext.mockRestore;
};

export const mockTeam = (team: Team) => {
  const teamContext = jest.spyOn(TeamContext, "useTeamState");
  teamContext.mockImplementation(() => team);
};

export * from "../test";
