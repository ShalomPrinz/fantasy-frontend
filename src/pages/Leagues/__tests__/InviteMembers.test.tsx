import { UserState } from "contexts";
import { mockUser, renderWithRouterAndUser } from "setupTests";

import InviteMembers from "../InviteMembers";

const renderComponent = (path?: string) =>
  renderWithRouterAndUser(<InviteMembers />, path);

describe("InviteMembers", () => {
  it("should render InviteMembers component if loading user", async () => {
    mockUser({ state: UserState.LOADING_USER });
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render InviteMembers component if no user is logged in", async () => {
    mockUser({ state: UserState.NO_LOGGED_USER });
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  describe("If User is logged in", () => {
    it("should render InviteMembers component if user is not a member of the league", async () => {
      mockUser({ state: UserState.LOGGED_USER });
      const { asFragment } = renderComponent();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
