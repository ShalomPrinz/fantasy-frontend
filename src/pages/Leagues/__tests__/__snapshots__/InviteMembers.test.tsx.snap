// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InviteMembers If User is logged in should render InviteMembers component if user is not a member of the league 1`] = `
<DocumentFragment>
  <h1
    class="text-center my-5 p-5 bg-danger text-white w-75 mx-auto"
  >
    You are not member of this league.
  </h1>
</DocumentFragment>
`;

exports[`InviteMembers should render InviteMembers component if loading user 1`] = `
<DocumentFragment>
  <h1
    class="text-center my-5 p-5 bg-info text-white w-75 mx-auto"
  >
    Loading...
  </h1>
</DocumentFragment>
`;

exports[`InviteMembers should render InviteMembers component if no user is logged in 1`] = `
<DocumentFragment>
  <h1
    class="text-center my-5 p-5 bg-danger text-white w-75 mx-auto"
  >
    Please Log In to invite members to your league
  </h1>
</DocumentFragment>
`;
